src/contracts/core/PaymentCoordinator.sol

```diff
@@ -0,0 +1,164 @@
+        merkleRoots[rootIndex].root = bytes32(0);
+    }
+
+    // @notice Permissioned function which allows withdrawal of EigenLayer's share of `token` from all received payments
+    function withdrawEigenlayerShare(IERC20 token, address recipient) external onlyRootPublisher{
+        uint256 amount = cumulativeEigenLayerTokeEarnings[token];
+        token.safeTransfer(recipient, amount);
+        cumulativeEigenLayerTokeEarnings[token] = 0;
+    }
+
+    /**
+    * @notice Called by a staker or operator to prove the inclusion of their earnings in a posted Merkle root and claim them.
+    * @param proof Merkle proof showing that a leaf containing `(msg.sender, amount)` was included in the `rootIndex`-th
+    * Merkle root posted for the `token`
+    * @param rootIndex Specifies the node inside the Merkle tree corresponding to the specified root, `merkleRoots[rootIndex].root`.
+    */
+    function proveAndClaimEarnings(
+        bytes memory proof,
+        uint256 rootIndex,
+        MerkleLeaf memory leaf
+    ) external{
+        require(leaf.amounts.length == leaf.tokens.length, "PaymentCoordinator.proveAndClaimEarnings: leaf amounts and tokens must be same length");
+        require(merkleRoots[rootIndex].confirmedAtBlockNumber < block.number, "PaymentCoordinator.proveAndClaimEarnings: Merkle root not yet confirmed");
+        bytes32 root = merkleRoots[rootIndex].root;
+        require(root != bytes32(0), "PaymentCoordinator.proveAndClaimEarnings: Merkle root is null");
+        bytes32 leafHash = keccak256(abi.encodePacked(leaf.recipient, keccak256(abi.encodePacked(leaf.tokens)), keccak256(abi.encodePacked(leaf.amounts)), leaf.index));
+        require(Merkle.verifyInclusionKeccak(proof, root, leafHash, leaf.index), "PaymentCoordinator.proveAndClaimEarnings: Invalid proof");
+
+        for(uint i = 0; i < leaf.amounts.length; i++) {
+            leaf.tokens[i].safeTransfer(leaf.recipient, leaf.amounts[i]);
+            cumulativeTokenAmountClaimedByRecipient[leaf.tokens[i]][leaf.recipient] += leaf.amounts[i];
+        }
+        emit PaymentClaimed(leaf);
+    }
+
+    function setRootPublisher(address _rootPublisher) external onlyOwner{
+        _setRootPublisher(_rootPublisher);
+    }
+
+    function setMerkleRootActivationDelay(uint256 _merkleRootActivationDelay) external onlyOwner{
+        _setMerkleRootActivationDelay(_merkleRootActivationDelay);
+    }
+
+    function setEigenLayerShareBIPS(uint256 _eigenlayerShareBips) external onlyOwner{
+        _setEigenLayerShareBIPS(_eigenlayerShareBips);
+    }
+
+
+    /// @notice Getter function for the length of the `merkleRoots` array
+    function merkleRootsLength() external view returns (uint256){
+        return merkleRoots.length;
+    }
+
+    function _setRootPublisher(address _rootPublisher) internal {
+        rootPublisher = _rootPublisher;
```

Comment:
there should probably be an associated `RootPublisherChanged` or similarly named event for this

Reply:
the event should emit the current (past) and new (updated) addresses

Reply:
added.